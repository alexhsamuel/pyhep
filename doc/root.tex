\section{\module{hep.root} -- Root interface}

\declaremodule{extension}{hep.root}
\modulesynopsis{Root interface}

The \module{hep.root} module provides access from Python to features in
Root, a framework for high energy physics data management and analysis.
The applicable components of Root are included with \pyhep and need not
be provided separately.

\subsection{Root files}

Root includes an implementation of histograms and ntuples (which are
called ``trees'' in Root).  It implements a file format for storing
these in structured disk files.  \pyhep provides read and write access
to histograms and trees stored in Root files.

The \function{open} function access a Root file:

\begin{funcdesc}{open}{path\optional{, mode="r", purge_cycles=1}}
 Create or open a Root file at \var{path}.  

 \var{mode} is the access mode, analogous to Python's built-in
 \function{file} and \function{open} functions.  If \code{"r"}, opens an
 existing file for reading.  If \code{"r+"}, \code{"w"}, \code{"a"}, or
 \code{"a+"}, opens an existing file for reading and writing, or creates
 the file if it doesn't exist.  If \code{"w+"}, creates a file or
 overwrites an existing file.

 If \var{purge_cycles} is set to a true value and the file is open for
 writing, the entire file will be purged when it is closed.  
\end{funcdesc}

The return value from \function{open} is a \class{File} object.  The
file remains open until it is deleted.  

In \pyhep, all paths in Root files are specified relative to the root
directory of the Root file.  The root directory itself is represented by
an empty string, and the path separator is a forward slash.  Note that
there is no notion of working directory or relative paths.

Root files also support revision cycles; see the Root documentation for
details.  Unless \function{root.open} is invoked with
\code{purge_cycles=0} (or the file is opened read-only), \pyhep
``purges'' the entire file when the file is closed.  This removes cycles
except the most recent for each object in the file, whether or not the
object was written or revised in the current session.

\pyhep stores additional information along with tables and histograms in
Root files.  This information, called \emph{metadata}, consists of
useful information about histograms and tables that is not handled
directly by the Root file format.  \pyhep stores metadata in
subdirectories named ``_pyhep_metadata'' in the Root file.  These
subdirectories and their contents may be removed safely without injuring
the primary data in the histograms and tables they are associated with,
but doing so will delete configuration and annotations accessible from
\pyhep.

\subsubsection{\class{File} objects}

A \class{File} object has the following methods and attributes:

\begin{memberdesc}{writeable}
 True if the file is open for writing as well as reading.
\end{memberdesc}

\begin{methoddesc}{listdir}{\optional{path}}
 Return the contents of the directory specified by \var{path}.  The
 return value is a sequence of \code{(name, title, class_name)} tuples,
 where \var{name} is the object's name, \var{title} is the object's
 title, and \var{class_name} is the object's class name.
\end{methoddesc}

\begin{methoddesc}{mkdir}{path\optional{, title}}
 Create a new directory at \var{path}.  Optionally, a title may be
 specified in the \var{title} argument.
\end{methoddesc}

\begin{methoddesc}{rmdir}{path}
 Remove the directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{save}{path, object}
 Store a histogram \var{object} at \var{path}.  One- and two-dimensional
 histograms are supported.  
\end{methoddesc}

\begin{methoddesc}{load}{path}
 Load a histogram or connect to an ntuple at \var{path}.  If \var{path}
 is a one- or two-dimensional histogram, returns a histogram object with
 its contents.  If \var{path} is a tree, returns a table object (see
 \module{hep.table}) accessing it.
\end{methoddesc}

\begin{methoddesc}{createTable}{path, schema\optional{, title,
separate_branches=0, branch_name="default"}}
 Create a table as a tree at \var{path} in the Root file, using
 columns from \var{schema}, a \class{hep.table.Schema} instance.  

 By default, column values is stored in a leaves grouped in a single
 branch, named "default".  Another name may be specified for the branch
 with the \var{branch_name} argument.  Alternately, if
 \var{separate_branches} is true, each leaf is placed in a separate
 branch, whose name is the column name.

 These column types are supported: \constant{"int8"},
 \constant{"int16"}, \constant{"int32"}, \constant{"float32"}, and
 \constant{"float64"}.
\end{methoddesc}

\begin{methoddesc}{remove}{path}
 Remove the tree or histogram at \var{path}.  Use \method{rmdir} to
 remove a directory.
\end{methoddesc}

\begin{methoddesc}{normpath}{path}
 Return the canonical form of absolute path \var{path} in the Root
 file.
\end{methoddesc}

\begin{methoddesc}{join}{*components}
 Join together \var{components} to form an absolute path.
\end{methoddesc}

\begin{methoddesc}{split}{path}
 Return \code{(dir, name)}, where \var{dir} is the portion of \var{path}
 up to the last directory separator, and \var{name} is the remainder.
\end{methoddesc}

