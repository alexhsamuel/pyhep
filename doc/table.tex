\section{\module{hep.table} --- Tables}

\declaremodule{extension}{hep.table}
\modulesynopsis{Tables.}

A \pyhep \emph{table} is a flat database table with columns of fixed
type and storage size.  Tables are append-only: new rows may be added to
the end of the table, but rows may not be inserted elsewhere, nor can
rows be modified or deleted.  

Each table has a fixed \emph{schema}, which determines the data items
stored in the table and their types.  The schema includes an unordered
set of \emph{columns}, each identified by a name.  Each column has a
fixed data type.  A schema may also include \emph{expressions}, which
are derived quantities which are computed for each row from values of
columns and other expressions for that row.  

The table's data is an ordered sequence of \emph{rows}, each of which
contains exactly one value for each column in the table's schema.

Multiple table implementations are provided.  Implementations provide
different interfaces for creating and opening tables.  However, objects
representing tables and rows satisfy the same protocol across all
implementations.


\subsection{Schema objects}

An instance of \class{hep.table.Schema} represents a table schema.  

\begin{memberdesc}{columns}
 \readonly A sequence of columns in the schema.  Each element is a
 \class{Column} instance.  The order of the columns is unspecified.
\end{memberdesc}

\begin{memberdesc}{expressions}
 \readonly A sequence of expressions in the schema.  Each element is
 expression object (see \module{hep.expr}).  The order of the
 expressions is unspecified.
\end{memberdesc}

\begin{methoddesc}{addColumn}{name, type\optional{, **attributes}}
 Add a column to the sequence.  \var{name} is a character string, and
 \var{type} is a column type (see below).  If additional keyword
 arguments are provided, they are added to the column as attributes.
 New columns should not be added to a schema after the schema has been
 used to create a table.
\end{methoddesc}

\begin{methoddesc}{addExpression}{name, expression\optional{, **attributes}}
 Add an expression to the sequence.  \var{name} is a character string,
 and \var{expression} is an expression as a string or expression object.
 If additional keyword arguments are provided, they are added to the
 expression object as attributes.
\end{methoddesc}

An instance of \class{hep.table.Column} describes a column in a schema.
Do not instantiate this class directly; instead, use the
\method{addColumn} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The column's name.
\end{memberdesc}

\begin{memberdesc}{type}
 \readonly The column's data type.
\end{memberdesc}

The column's data type is specified by one of these names:
\begin{tableiii}{ccc}{code}{name}{C type}{Python Type}
  \lineiii{"int8"}{\code{signed char}}{\code{int}}
  \lineiii{"int16"}{\code{signed short}}{\code{int}}
  \lineiii{"int32"}{\code{signed int}}{\code{int}}
  \lineiii{"float32"}{\code{float}}{\code{float}}
  \lineiii{"float64"}{\code{double}}{\code{float}}
\end{tableiii}


An instance of \class{hep.table.Expression} describes a column in a
schema.  Do not instantiate this class directly; instead, use the
\method{addExpression} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The expressions's name.
\end{memberdesc}

\begin{memberdesc}{expression}
 \readonly An \class{Expression} instance; see \module{hep.expr}.
\end{memberdesc}

\begin{memberdesc}{formula}
 \readonly A string representation of the expression.
\end{memberdesc}



\subsection{Table objects}

A table object represents an in-memory table or a connection to an
externally-stored table.  In the former case, the table is erased when
the table object is deleted.  In the latter case, the connection is
closed but the table persists when the table object is deleted.

A table object satisfies the Python read-only sequence protocol.  The
sequence elements are the rows of the sequence.  Thus, the
\function{len} function returns the number of rows in the table, and
the \method{__getitem__} method extracts a row by its position in the
sequence. 

In addition, these methods are provided:

\begin{methoddesc}{append}{row\optional{, **kw_args}}
 (Read-write tables only.)  Appends a row to the end of the table.  The
 \var{row} must is a dictionary or other mapping object which maps
 column names to values.  Additional column values may be specified as
 keyword arguments.  If a column is specified both in \var{row} and in
 \var{**kw_args}, the value from latter is used.  A value must be
 specified for each column, and no other names may be specified.
 Returns the row index of the new row.
\end{methoddesc}

\begin{methoddesc}{__iter__}{}
 Returns an iterator over all rows in the table.
\end{methoddesc}

\begin{memberdesc}{rows}
 An interator over all rows in the table.
\end{memberdesc}

\begin{methoddesc}{select}{expr}
 Returns an iterator over rows in the table for which expression
 \var{expr} is true.  \var{expr} may be a string expression formula or
 an expression object.
\end{methoddesc}


\subsection{Row objects}

A row object may represent a row in a table, or a row that is compatible
with a table's schema (but not one of the rows in that table).

The row object satisfies the Python map protocol.  Each key is the name
of a column in the table's schema.  The corresponding map value is the
row's value for that column.  Row objects should be treated as
read-only; it is not possible to modify a table row by changing values
in a row object.

In addition, a row object provides these members and methods:

\begin{methoddesc}{asDict}{}
 Return a new Python dictionary of the column names and corresponding
 values in the row.
\end{methoddesc}

\begin{memberdesc}{index}
 \readonly The index of the row in the table.  The value of this
 attribute is undefined if the row was obtained from the table's
 \method{newRow} method and has not yet been passed to \method{append}.
\end{memberdesc}

\begin{memberdesc}{table}
 \readonly The table to which the row belongs.
\end{memberdesc}


\subsection{Standard implementation}

\pyhep's standard table implementation, in \module{hep.table}, uses a
simple, binary, on-disk representation of the table's data.  Table rows
are automatically loaded from disk as needed.

The \module{hep.table} module provides functions \function{create} and
\function{open} to create and open new tables, respectively.

\begin{funcdesc}{create}{filename, schema}
 Create a new table.  The table is stored in a file named by
 \var{filename}, which is created or overwritten.  The \var{schema}
 argument is a map from column names to \class{Column} instances,
 specifying the columns in the table.

 The return value is a table object, which is opened to the new, empty
 in write mode.
\end{funcdesc}

\begin{funcdesc}{open}{filename\optional{, mode="r"}}
 Open an existing table stored in the file named by \var{filename}.  The
 \var{mode} argument specifies the mode in which to open the table:
 \constant{"r"} to open the table in read-only mode, or \constant{"w"}
 to open the table in read-write mode.

 The return value is a table object.
\end{funcdesc}

