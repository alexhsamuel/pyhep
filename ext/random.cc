//----------------------------------------------------------------------
//
// random.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cstddef>
#include <sys/time.h>

#include "random.hh"

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

namespace {

// Constants for the first congruential generator.
const static long m1 = 2147483563;
const static long a1 = 40014;
const static long q1 = 53668;
const static long r1 = 12211;

// Constants for the second congruential generator.
const static long m2 = 2147483399;
const static long a2 = 40692;
const static long q2 = 52774;
const static long r2 = 3791;

//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline float
min(float v1,
    float v2)
{
  return (v1 < v2) ? v1 : v2;
}


}  // anonymous namespace

//----------------------------------------------------------------------
// methods
//----------------------------------------------------------------------

ShuffledLEcuyerRandom::ShuffledLEcuyerRandom(long seed)
{
  initialize(seed);
}


void
ShuffledLEcuyerRandom::initialize(long seed)
{
  // If no seed was specified, construct one from the system time.
  if (seed == 0) {
    struct timeval now;
    int result = gettimeofday(&now, NULL);
    assert(result == 0);
    seed = now.tv_sec ^ now.tv_usec;
  }

  // Initialize.
  assert(seed > 0);
  state1_ = seed;
  state2_ = seed;

  // Load the shuffle table (after eight warm-ups).
  for (long j = table_size_ + 7; j >= 0; j--) {
    long k = state1_ / q1;
    state1_ = a1 * (state1_ - k * q1) - k * r1;
    if (state1_ < 0)
      state1_ += m1;
    if (j < table_size_)
      shuffle_table_[j] = state1_;
  }
  last_ = shuffle_table_[0];
}


float
ShuffledLEcuyerRandom::random()
{
  long k;
  static float inv_m1 = 1.0 / m1;

  // Next value from the first congruential generator.
  k = state1_ / q1;
  state1_ = a1 * (state1_ - k * q1) - k * r1;
  if (state1_ < 0)
    state1_ += m1;

  // Next value from the second congruential generator.
  k = state2_ / q2;
  state2_ = a2 * (state2_ - k * q2) - k * r2;
  if (state2_ < 0)
    state2_ += m2;

  // Combine and shuffle.
  int j = last_ / (1 + (m1 - 1) / table_size_);
  last_ = shuffle_table_[j] - state2_;
  shuffle_table_[j] = state1_;
  if (last_ < 1)
    last_ += m1 - 1;

  // Make sure we never return 1.0.
  return min(last_ * inv_m1, 1.0 - 1.2e-7);
}


