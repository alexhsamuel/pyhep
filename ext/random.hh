//----------------------------------------------------------------------
//
// random.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

#ifndef __RANDOM_HH__
#define __RANDOM_HH__

//----------------------------------------------------------------------
// classes
//------------------------------------------------------------------------

/* Shuffled L'Ecuyer random number generator.

   Generator for uniform deviates based on 

     L'Ecuyer, P. 1988, Communications of the ACM, vol. 31, pp. 742-774

   as described and implemented as the 'ran2' function in

     Press, William H. et al., Numerical Recipes in C, Cambridge
     University Press, 1988, p. 282.

   This generator is based on two multiplicative congruential generators
   with factors 'a1' and 'a2' and moduli 'm1' and 'm2, with
   multiplication performed with Schrage's algorith.  The values from
   the two sequences are combined, and the sequence is shuffled using a
   table of 'tab_size'.  

   The period of the generator is about 2.3 x 10^18.

*/
class ShuffledLEcuyerRandom
{
public:

  /* Create a new generator.

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  ShuffledLEcuyerRandom(long seed=0);
  
  /* Initialize the generator. 

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  void initialize(long seed=0);

  /* Return a uniform deviate between 0 inclusive and 1 exclusive.  */
  float random();

private:

  // Forbid copies.
  ShuffledLEcuyerRandom(const ShuffledLEcuyerRandom&);

  // Shuffle table size.
  const static int table_size_ = 32;

  // State variables for the congruential generators.
  long state1_;
  long state2_;

  // State of the combined generator; used for shuffling.
  long last_;

  // The shuffle table.
  long shuffle_table_[table_size_];

};

//----------------------------------------------------------------------

#endif  // #ifndef __RANDOM_HH__
