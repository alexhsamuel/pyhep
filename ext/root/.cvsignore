root.py
root_wrap.cc
root_wrap.doc
*Dict.cc
*Dict.h
*.pyc
depend.mk
ROOTSYS
