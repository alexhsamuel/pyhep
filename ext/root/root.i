//------------------------------------------------------------*- c++ -*-
//
// root.i
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

// SWIG interface file for Root libraries.

// -*- c++ -*-

%module root
%{
#include "TROOT.h"
#include "TKey.h"
#include "TDirectory.h"
#include "TFile.h"
#include "TLeaf.h"
#include "TLeafD.h"
#include "TLeafF.h"
#include "TLeafI.h"
#include "TList.h"
#include "TObject.h"
#include "TTree.h"
#include "TH1.h"
#include "TH2.h"
#include "TH3.h"
#include "TAxis.h"

class IndexError 
{
public:
  IndexError(int index) : index_(index) {};
  const int index_;
};

class StopIteration {};

%}

%inline %{
long 
getPointerValue(void* ptr) 
{
   return (long) ptr;
}
%}

%except(python) {
  try {
    $function
  }
  catch (IndexError e) {
    PyErr_Format(PyExc_IndexError, "%d", e.index_);
    return NULL;
  }
  catch (StopIteration) {
    PyErr_SetString(PyExc_StopIteration, "end of iterator");
    return NULL;
  }
}


class TObject
{
public:

  bool IsZombie() const;
  virtual const char* ClassName() const;
  virtual const char* GetName() const;
  virtual const char* GetTitle() const;
  virtual bool InheritsFrom(const char *classname) const;
  virtual void Print(const char* option) const;
  virtual int Write(const char* name=0, int option=0, int bufsize=0);

};


class TNamed
  : public TObject
{
public:

  virtual void SetName(const char* name);
  virtual void SetTitle(const char* title);

};


%pragma no_default

class TIterator
  : public TObject
{
public:

  virtual TObject* Next();

};

%addmethods TIterator {
  TIterator* __iter__() 
  {
    return self;
  };


  TObject* next()
  {
    TObject* object = self->Next();
    if (object == NULL)
      throw StopIteration();
    else
      return object;
  };
};



class TClass
  : public TObject
{
public:

  bool InheritsFrom(const char* cl) const;

};


class TKey
  : public TNamed
{
public:

  const char* GetClassName() const;
  short GetCycle() const;
  virtual TClass* IsA() const;

};


class TDirectory
  : public TNamed
{
public:

  virtual bool cd();
  virtual void Delete(const char* namecycle);
  virtual TObject* Get(const char* namecycle);
  TFile* GetFile() const;
  TKey* GetKey(const char* name, short cycle=9999);
  TList* GetList() const;
  TList* GetListOfKeys() const;
  TObject* GetMother() const;
  const char* GetPath() const;
  virtual TDirectory* mkdir(const char* name, const char* title);
  void Purge(short nKeep);
  
};


class TFile
  : public TDirectory
{
public:

  TFile(const char* fname, 
	char* option="", 
	const char* ftitle="", 
	int compress=1);

  virtual bool cd(const char *path = 0);
  virtual void Close(char* option="");
  virtual int Write(const char* name=0, int opt=0, int bufsiz=0);

};


class TList
  : public TObject
{
public:

  virtual int GetSize() const;
  virtual TObject* At(int idx) const;

};


%addmethods TList {
  int __len__() 
  {
    return self->GetSize();
  };


  TObject* __getitem__(int idx) 
  {
    if (idx < 0 || idx >= self->GetSize()) 
      throw IndexError(idx);
    else
      return self->At(idx);
  };
};



class TTree
  : public TObject
{
public:

  virtual int GetEntry(int entry=0, int getall=0);
  virtual double GetEntries() const;
  virtual TIterator* GetIteratorOnAllLeaves();
  virtual void Print();

};


class TLeaf
  : public TObject
{
public:
  
  virtual void SetAddress(void* add=0);

};

%addmethods TLeaf {
  void SetAddressInt(int add=0) 
  {
    self->SetAddress((void*) add);
  };
};


class TLeafD
  : public TLeaf
{
};


class TLeafF
  : public TLeaf
{
};


class TLeafI
  : public TLeaf
{
};


class TH1
  : public TNamed
{
public:

  virtual TAxis* GetXaxis();
  virtual TAxis* GetYaxis();
  virtual TAxis* GetZaxis();
  virtual double GetBinContent(int bin);
  virtual double GetBinError(int bin);
  virtual double GetEntries();
  virtual void SetBinContent(int bin, double content);
  virtual void SetBinError(int bin, double error);
  virtual void SetEntries(double entries);

};


class TH1F 
  : public TH1
{
public:

  TH1F(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1D
  : public TH1
{
public:

  TH1D(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1C
  : public TH1
{
public:

  TH1C(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1S
  : public TH1
{
public:

  TH1S(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH2
  : public TH1
{
public:

  virtual double GetBinContent(int binx, int biny);
  virtual double GetBinError(int binx, int biny);
  virtual void SetBinContent(int binx, int biny, double content);
  virtual void SetBinError(int binx, int biny, double error);

};


class TH2C
  : public TH2
{
public:

  TH2C(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2D
  : public TH2
{
public:

  TH2D(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2F
  : public TH2
{
public:

  TH2F(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2S
  : public TH2
{
public:

  TH2S(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH3
  : public TH1
{
};


class TAxis
  : public TObject
{
public:

  int GetNbins();
  double GetXmin();
  double GetXmax();

};
