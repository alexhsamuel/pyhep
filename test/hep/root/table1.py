#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.root
import hep.table
from   random import random
from   hep.test import compare, assert_

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

schema = hep.table.Schema()
schema.addColumn("index", "int32")
schema.addColumn("value32", "float32")
schema.addColumn("value64", "float64")

root_file = hep.root.create("table1.root")
table = hep.root.createTable("table", root_file, schema)
values = []
for i in xrange(0, 100):
    value = random()
    values.append(value)
    table.append(index=i, value32=value, value64=value)
del root_file, table

table = hep.root.open("table1.root")["table"]
compare(len(table), 100)
i = 0
for row in table.rows:
    compare(row["index"], i)
    compare(row["value32"], values[i], precision=1e-6)
    compare(row["value64"], values[i], precision=1e-9)
    i += 1

