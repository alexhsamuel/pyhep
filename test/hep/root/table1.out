Traceback (most recent call last):
  File "table1.py", line 28, in ?
    table = hep.root.open("table1.root")["table"]
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 667, in open
    return file.root
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 541, in <lambda>
    root = property(lambda self:
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 65, in __init__
    for tkey in [ _cast(tlist[i]) for i in range(len(tlist)) ]:
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 600, in _cast
    parts = object.this.split("_")
AttributeError: 'PySwigObject' object has no attribute 'split'
